Support for sharing extend metadata between grpc and http

CrazyHZM opened the pull request:
Signed-off-by: crazyhzm <crazyhzm@gmail.com>

# Description

<!--
Please explain the changes you've made.
-->

## Issue reference

<!--
We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.
-->

Please reference the issue this PR will close: #4941

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

daixiang0 commented on pkg/metadata/metadata_store.go:
```diff
@@ -0,0 +1,41 @@
+}
+
+// MetadataGet get metadata of sidecar.
+func (m *DefaultStore) MetadataGet() (map[string]string, error) {
```
Always nil error? Could we remove it?

CrazyHZM replied:
If the follow-up is no longer in-memory storage, it may no longer be a nil error.

tmacam replied:
Is there any plan metadata stored in something else rather than in-memory storage?

saber-wang commented on pkg/grpc/api.go:
```diff
@@ -1470,22 +1474,17 @@ func (a *api) SetActorRuntime(actor actors.Actors) {
 }
 
 func (a *api) GetMetadata(ctx context.Context, in *emptypb.Empty) (*runtimev1pb.GetMetadataResponse, error) {
```
The returned content is missing `ID`,`ActiveActorsCount`

saber-wang replied:
There are many errors in the previous implementation of  `GetMetadata` , which should be modified by referring to HTTP API

CrazyHZM replied:
Already support.

saber-wang commented on pkg/grpc/api.go:
```diff
@@ -1470,22 +1474,17 @@ func (a *api) SetActorRuntime(actor actors.Actors) {
-		return true
-	})
 	temp[daprRuntimeVersionKey] = a.daprRunTimeVersion
 	registeredComponents := make([]*runtimev1pb.RegisteredComponents, 0, len(a.components))
```
`a.components' has no value, because the parameter `getComponentsFn` is not assigned to `components` when `NewAPI`

CrazyHZM replied:
This has been solved.

CrazyHZM commented:
@yaron2 This failing test has nothing to do with this PR.

yaron2 commented:
> @yaron2 This failing test has nothing to do with this PR.

That's correct.

tmacam commented on pkg/grpc/api.go:
```diff
@@ -1472,27 +1473,13 @@ func (a *api) SetActorRuntime(actor actors.Actors) {
 }
 
 func (a *api) GetMetadata(ctx context.Context, in *emptypb.Empty) (*runtimev1pb.GetMetadataResponse, error) {
-	extendedMetadata := make(map[string]string)
+	temp, _ := a.extendedMetadata.MetadataGet()
```
Could we be a bit more descriptive than `temp`?

berndverst commented:
This PR is out of date. #5018 already addressed some of the concerns, and #5052 aims to address the remaining concerns.

@tmacam can you please review this PR to check that it contains the changes you were trying to make? I think it makes sense to close #5052 in favor of this PR (FYI @artursouza , @yaron2 )

@CrazyHZM please prioritize getting this PR updated. Thanks!

CrazyHZM commented:
> This PR is out of date. #5018 already addressed some of the concerns, and #5052 aims to address the remaining concerns.
> 
> @tmacam can you please review this PR to check that it contains the changes you were trying to make? I think it makes sense to close #5052 in favor of this PR (FYI @artursouza , @yaron2 )
> 
> @CrazyHZM please prioritize getting this PR updated. Thanks!

@berndverst 
Of course, this pr is being continuously updated.

tmacam commented on pkg/grpc/api_test.go:
```diff
@@ -28,6 +28,8 @@ import (
 
 	"github.com/dapr/components-contrib/lock"
 
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Do we really need this dependency just to define some mocks down for test?

berndverst replied:
This import should be ok.

daixiang0 commented on pkg/grpc/api.go:
```diff
@@ -1472,27 +1473,13 @@ func (a *api) SetActorRuntime(actor actors.Actors) {
 
 func (a *api) GetMetadata(ctx context.Context, in *emptypb.Empty) (*runtimev1pb.GetMetadataResponse, error) {
-	extendedMetadata := make(map[string]string)
+	extendedMetadata, _ := a.extendedMetadata.MetadataGet()
```
Please handle error.

CrazyHZM commented:
@yaron2 @artursouza 
Please take care of this PR as soon as possible, I have dealt with the conflict very many times and have already solved the problem with the previously merged PR at the beginning.

tmacam commented on pkg/runtime/runtime.go:
```diff
@@ -1331,9 +1340,15 @@ func (a *DaprRuntime) getGRPCAPI() grpc.API {
 	)
 }
 
+func (a *DaprRuntime) storeMetadata() {
```
nitpicking: this method has a generic "storeMetadata" name when in reality it initializing DaprRuntimeVersion within the the extended metadata of the Metedata API -- this is too "meta" and abstract ;)  I would make the name a bit more more descriptive like "initializeExtendedMetadata".

tmacam commented on pkg/runtime/runtime.go:
```diff
@@ -466,6 +471,9 @@ func (a *DaprRuntime) initRuntime(opts *runtimeOpts) error {
 	// Start proxy
 	a.initProxy()
 
+	// Store dapr runtime metadata
```
nitpicking: s/metadata/within the extended metadata/

tmacam reviewed (commented):
LGTM aside some small nitpicking.

I sense there will be some conflict merging your files - seems some changes you are reproducing are already in the code.

mukundansundar commented on pkg/http/api.go:
```diff
@@ -1880,7 +1870,7 @@ func getOrDefaultCapabilites(dict map[string][]string, key string) []string {
 	key := fmt.Sprintf("%v", reqCtx.UserValue("key"))
 	body := reqCtx.PostBody()
-	a.extendedMetadata.Store(key, string(body))
+	_ = a.extendedMetadata.MetadataSet(key, string(body))
```
Please handle error properly instead of swallowing it here? I can see that it's always returning nil for now though... So why do we need it, is it for some future change?

mukundansundar commented on pkg/metadata/metadata_store.go:
```diff
@@ -0,0 +1,45 @@
+	m.metadata.Range(func(key, value interface{}) bool {
+		temp[key.(string)] = value.(string)
+
+		return true
```
return true, nil?

CrazyHZM replied:
Here is just a simple copy.

artursouza commented on pkg/metadata/metadata_store.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+// DefaultStore is a default implementation of Store.
+type DefaultStore struct {
```
What is the advantage of creating a new type?

CrazyHZM replied:
It is mainly used for runtime metadata storage, and is only stored in memory. In the future, runtime metadata may also be stored in other storage components.

daixiang0 commented:
/ot-to-test

yaron2 commented:
@CrazyHZM can you please resolve the conflicts?

CrazyHZM commented:
> 

Done @yaron2

CrazyHZM commented:
There are already too many conflicts in this PR. I will close this pr first and look at the problem again.
